src/Microsoft.AspNetCore.Mvc.Razor/RazorPage.cs

```diff
@@ -353,7 +340,10 @@ private static void WriteTo(TextWriter writer, HtmlEncoder encoder, string value
         {
             if (!string.IsNullOrEmpty(value))
             {
-                encoder.Encode(writer, value);
+                // Perf: Encode write away instead of writing it character-by-character.
+                // character-by-character isn't efficient when using a writer backed by a ViewBuffer.
+                var encoded = encoder.Encode(value);
+                writer.Write(encoded);
```

Comment:
Unfortunately both `ViewContext.Writer` and the `HtmlEncoder` are available in user code. So even if we avoid `Encode(TextWriter, string)` completely, users can shoot themselves in both feet simultaneously. I'm also not sure the `ViewBuffer` would need to detect matching encoders. How about instead overriding `Write(char)` in `ViewBufferTextWriter` to use a `bool _isUsingSetAsideBuffer` and a `StringBuilder _setAsideBuffer`, where the `_setAsideBuffer` is lazily initialized but reused.